### ISSUE TITLE
We can add a new api using operator-sdk 0.0.6 - must run command in project root dir: stat ./build/Dockerfile: no such file or directory

### ISSUE THREAD

**Issue by cmoulliard** [operator-framework/operator-sdk#655]
### Command issued

```
operator-sdk add api --api-version=component.k8s.io/v1alpha1 --kind=Export
2018/10/23 17:42:55 must run command in project root dir: stat ./build/Dockerfile: no such file or directory
```

### Tree of the local project

```
 tree -I vendor
.
├── Gopkg.lock
├── Gopkg.toml
├── Makefile
├── README.md
├── cmd
│   ├── component-operator
│   │   └── main.go
│   └── sd
│       └── sd.go
├── config
│   └── config.yaml
├── deploy
│   ├── cr.yaml
│   ├── crd.yaml
│   ├── operator.yaml
│   ├── rbac.yaml
│   └── sa.yaml
├── examples
│   ├── component-crud.yml
│   ├── component-link.yml
│   ├── component-route.yml
│   ├── component-service.yml
│   ├── component-vertx.yml
│   └── component.yml
├── hack
│   ├── k8s-api.sh
│   └── steps.md
├── pkg
│   ├── apis
│   │   └── component
│   │       └── v1alpha1
│   │           ├── doc.go
│   │           ├── register.go
│   │           ├── types.go
│   │           └── zz_generated.deepcopy.go
│   ├── client
│   │   └── cmd
│   │       ├── create.go
│   │       ├── root.go
│   │       └── version.go
│   ├── integration
│   │   ├── enum_test.go
│   │   ├── enum_weekday.go
│   │   └── json_test.go
│   ├── stub
│   │   ├── handler.go
│   │   └── pipeline
│   │       ├── generic
│   │       │   ├── pvc.go
│   │       │   ├── route.go
│   │       │   └── service.go
│   │       ├── innerloop
│   │       │   ├── deploymentconfig.go
│   │       │   ├── images.go
│   │       │   ├── install.go
│   │       │   └── resources.go
│   │       ├── link
│   │       │   └── link.go
│   │       ├── servicecatalog
│   │       │   └── install.go
│   │       └── step.go
│   └── util
│       ├── kubernetes
│       │   ├── config.go
│       │   ├── convert.go
│       │   └── namespace.go
│       ├── oc
│       │   └── oc.go
│       ├── openshift
│       │   └── register.go
│       └── template
│           ├── assets.go
│           ├── assets_vfsdata.go
│           ├── doc.go
│           ├── template.go
│           └── tmpl
│               ├── innerloop
│               │   ├── deploymentconfig
│               │   ├── imagestream
│               │   ├── pvc
│               │   ├── route
│               │   └── service
│               └── servicecatalog
│                   ├── servicebinding
│                   └── serviceinstance
├── tmp
│   ├── _output
│   │   └── bin
│   │       ├── component-operator
│   │       ├── sb
│   │       ├── sd
│   │       └── spring-boot-operator
│   ├── build
│   │   ├── Dockerfile
│   │   ├── build.sh
│   │   ├── go-test.sh
│   │   └── test-framework
│   │       └── Dockerfile
│   └── codegen
│       ├── boilerplate.go.txt
│       └── update-generated.sh
└── version
    └── version.go

```

**Environment**
* operator-sdk version: 0.0.6+git

* Kubernetes version information:

  Insert output of `kubectl version` here

* Kubernetes cluster kind: 

```
oc version
oc v3.11.0+0cbc58b
kubernetes v1.11.0+d4cacc0
features: Basic-Auth

Server https://192.168.99.50:8443
kubernetes v1.11.0+d4cacc0
```

* Are you writing your operator in ansible or go? go

**Comment by cmoulliard** [comment]
A temporary workaround is to create a symbolic link

```
ln -s ./tmp/build build           
```

But, now, I get such problems

```
operator-sdk add api --api-version=component.k8s.io/v1alpha1 --kind=Export
Create pkg/apis/component/v1alpha1/export_types.go
Create pkg/apis/addtoscheme_component_v1alpha1.go
Create deploy/crds/component_v1alpha1_export_cr.yaml
Create deploy/crds/component_v1alpha1_export_crd.yaml
2018/10/23 18:02:55 failed to update the RBAC manifest for the resource (component.k8s.io/v1alpha1, Export): failed to read role manifest /Users/dabou/Code/go-workspace/src/github.com/snowdrop/component-operator/deploy/role.yaml: open /Users/dabou/Code/go-workspace/src/github.com/snowdrop/component-operator/deploy/role.yaml: no such file or directory
```

**Comment by mhrivnak** [comment]
It looks like you're running from `master`, which broke compatibility after 0.0.7. I think the reported version of "0.0.6+git" is misleading, and was just an oversight that it didn't get updated in-code.

I suggest using the 0.0.7 release for now. When `master` gets released as 0.1.0, there will be a guide to dealing with the broken compatibility.

**Comment by cmoulliard** [comment]
Can we add an `api` or `operator` with version `0.0.7` or the only way is to create a new project ? @mhrivnak

**Comment by hasbro17** [comment]
If you build the operator-sdk from the master you should be able to to use the commands `add api` and `add controller` but running them in your existing project might cause some issues.

We're working on a migration guide right now in which you essentially have to create a new project and copy your reconcile code and other files over because there are some differences in the project layout post `v0.0.7`. We'll have the guide finished and `v0.1.0` release out soon.
https://github.com/operator-framework/operator-sdk/pull/653

It is possible to first fix up your project layout first and then run `add api` and `add controller` to migrate your project in place but we haven't documented the steps for that yet, so that's not currently supported. Plus only running `add api` would still cause issues since that type wouldn't be registered with the sdk scheme.

We're adding a doc to highlight the layout differences but you can compare the sample memcached-operator in the meanwhile to see the differences:
[v0.0.7-memcached-operator](https://github.com/operator-framework/operator-sdk-samples/tree/aa15bd278eec0959595e0a0a7282a26055d7f9d6/memcached-operator)
[v0.0.7+git-memcached-operator](https://github.com/operator-framework/operator-sdk-samples/tree/master/memcached-operator)
https://github.com/operator-framework/operator-sdk/blob/master/doc/project_layout.md

**Comment by cmoulliard** [comment]
We still have the issue reported within the title of this ticket using the `operator-sdk` version `0.1.0`

```
operator-sdk add api --api-version=component.k8s.io/v1alpha1 --kind=Component
2018/11/16 09:25:03 must run command in project root dir: stat ./build/Dockerfile: no such file or directory
```

The workaround is to copy the `Dockerfile` under the `./build` directory

**Question**  : Why the command adding a new API is looking about the Dockerfile which is only needed when we build the docker image if the operator ?

**Comment by LiliC** [comment]
@cmoulliard This is because the `new` commands creates the `build/Dockerfile` and there is a check to make sure that while running the `add` command the user is in the root directory, that check looks for the Dockerfile. This is because it is also used during other commands such as `build`. 

This check is used in ansible as well and that is the only common file hence why we look for the `Dockerfile`. I think the error msg can be cleared up a bit, but we can't check for any other file right now.

@cmoulliard I am curious why do you not have a Dockerfile there in the first place? And what would you like the outcome to be?

**Comment by cmoulliard** [comment]
> Dockerfile there in the first place

Just because I have created a `circleci` job doing the work to compile the go binary and next to create our operator's image - See [here](https://github.com/snowdrop/component-operator/blob/master/.circleci/config.yml#L63)

**Comment by cmoulliard** [comment]
> And what would you like the outcome to be?

A warning message explaining that the Dockerfile is not defined under the following `./build/` directory and then the `operator-sdk build` could fail. TBH, I would prefer that no message is displayed when we generate the code of a new `API` as that will confuse the developers

**Comment by joelanford** [comment]
One option would be that we change our logic for how we decide we're in a project root directory. Right now, we check `./build/Dockerfile` for all operator types, but we could instead just check that we have a valid operator type.

Our `GetOperatorType()` function currently assumes we're in a project root, but there's no reason it has to. Its logic could be something like:

```go
func GetOperatorType() OperatorType {
    if IsGoOperator() {
        return OperatorTypeGo
    }
    if IsAnsibleOperator() {
        return OperatorTypeAnsible
    }
    return OperatorTypeUnknown
}
```

Then `MustInProjectRoot` would look something like:

```go
func MustInProjectRoot() {
    if GetOperatorType() == OperatorTypeUnknown {
        log.Fatal("must run command in project root dir: could not identify operator type")
    }
}
```

With this type of check, we would no longer need to rely on a single file that's present in every type of operator for an operation that doesn't depend on that file.

**Comment by LiliC** [comment]
@joelanford Overall that sounds like a good option. For the `IsGoOperator` we could just keep whatever we have currently in the `GetOperatorType`. What would the `IsAnsibleOperator` look like then? cc @shawn-hurley for any further ideas.

**Comment by LiliC** [comment]
@shawn-hurley Agreed, but that check should not be called `MustInProjectRoot` but maybe `MustHaveOperatorSDKStructure` or something, if you get what I am saying? And we should adjust the error msg to reflect that.

**Comment by shawn-hurley** [comment]
Fair, I am ok with changing the error message and function name, but what would it be? wouldn't it be just as nebulous as asking to be in the project root? I run this command from say <root>/pkg and we say "Must have the correct operator sdk structure" but they do they are just in the wrong directory? 

maybe a combo message but that gets wordy, I do think that we probably want to remove the error message from that message. What about "Expected correct structure and root directory to have ./build/Dockerfile"? This is confusing though IMO. I don't know what the right thing is to do with the error message.

**Comment by LiliC** [comment]
@shawn-hurley Yes I tried to come up with something as well, but failed to think of something that makes sense. :D But that sgtm, rephrased a bit. :)

@cmoulliard since you opened the original issue about this, would this make more sense to you:

> "Expected correct project structure and root directory to have ./build/Dockerfile. Double check that the ./build/Dockerfile is present and the command is run from root directory."

**Comment by cmoulliard** [comment]
> would this make more sense to you

The wording proposed is clear to me ;-) I repeat nevertheless that this checks is completely irrelevant when we add a new API and which is the purpose of my ticket @LiliC

**Comment by LiliC** [comment]
@cmoulliard I understand your concern, the problem is that is the only way we can check for the correct project structure by making sure the files we just generated are there.

I guess we could add a flag `--force` or `--skip-checks` and add the files regardles of the checks? WDYT @joelanford @shawn-hurley

**Comment by cmoulliard** [comment]
My reasoning is perhaps naive but when we execute the command `operator-sdk add api`, then the following resources are created under either `deploy` or `pkg/apis` folders

```
// Types is the input needed to generate a pkg/apis/<group>/<version>/<kind>_types.go file
// AddToScheme is the input needed to generate an pkg/apis/addtoscheme_<group>_<version>.go file
// Register is the input needed to generate a pkg/apis/<group>/<version>/register.go file
// Doc is the input needed to generate a pkg/apis/<group>/<version>/doc.go file
// Cr is the input needed to generate a deploy/crds/<group>_<version>_<kind>_cr.yaml file
// Crd is the input needed to generate a deploy/crds/<group>_<version>_<kind>_crd.yaml file

// append rbac rule to deploy/role.yaml

// K8sCodegen performs deepcopy code-generation for all custom resources under pkg/apis	
```

So, the only paths to be checked/verified are  `./deploy` and `./pkg/apis` - make sense ? @shawn-hurley

**Comment by joelanford** [comment]
@cmoulliard Out of curiosity, is there a reason you chose to move your Dockerfile to `./deploy/docker/Dockerfile`? If you moved it back to `./build/Dockerfile` and updated your circleci build config to point to that instead, would that solve this particular problem? Would something else in your project break?

@shawn-hurley How did we decide on `./build/Dockerfile` as the definitive marker for being in the project root? Was it just a matter of convenience (i.e. this file happens to exist for all project types)? Could we have just as easily chosen `./deploy/operator.yaml`? I do think the CLI should be opinionated, but perhaps it should only be as opinionated as necessary to perform the requested action.

**Comment by shawn-hurley** [comment]
> @shawn-hurley How did we decide on ./build/Dockerfile as the definitive marker for being in the project root? Was it just a matter of convenience (i.e. this file happens to exist for all project types)? Could we have just as easily chosen ./deploy/operator.yaml? I do think the CLI should be opinionated, but perhaps it should only be as opinionated as necessary to perform the requested action.

Yes, it is it a file that is available for all operators and is needed by the sdk to do other actions. `deploy/operator.yaml` can be changed I believe and no other commands will break (this might not be a valid assumption i.e `test`). But all this means is that you will need a check for project root for each command. If we want to add this complexity at this point that is a trade-off that should be discussed. I don't see the value for the complexity IMO. 

we could do what you said above, but that means that we need to determine if `Helm` operator or `Ansible` operator. We had discussed that the `Roles` and `Charts` directories would signal this, but does that solve the problem? I still think you eventually get this same complaint about a different file.

Just my two cents

**Comment by joelanford** [comment]
For the operator-specific check, I was thinking `cmd/manager/main.go`, './roles/`, and `./helm-charts/`, respectively, for go, ansible, and helm.

But that is a good point about getting a future complaint about one of those files, since a user could easily move `./roles/` to `./ansible-roles/` and modify other files accordingly to make things work. So you've convinced me that that approach would just be shifting the goalposts.

`operator-sdk add api` only cares about `./deploy/` and `./pkg/apis/`, so there's an argument to be made about checking that just those directories exist. But again, what stops a user from modifying those locations and customizing their layout? If we allow some modifications to the project layout but not others, that would increase complexity of the CLI and lead to more confusion from users.

So, having settled on that, back to a better error message...

Maybe:
```go
fmt.Errorf("must run command in project root dir: project structure requires ./build/Dockerfile: %s", err)
```

That way the underlying error of checking that file still gets propagated to the user (in case it's something other than not found).

**Comment by hasbro17** [comment]
@cmoulliard Even if we ignore `build/Dockerfile` during `add api` and check something else to ensure project root, you'll still run into a problem executing `operator-sdk build` which relies on `build/Dockerfile`.
https://github.com/operator-framework/operator-sdk/blob/master/commands/operator-sdk/cmd/build.go#L169
I assume you're bypassing the `build` command and just building manually.

But as @shawn-hurley and @joelanford pointed out we need to have some expectation of project structure even if it's not related to the command. Otherwise it opens the door to what users are allowed to move around and still expect the SDK to work.
I would recommend making a dummy `build/Dockerfile` if you don't need one in that location, or symlinking it to `deploy/docker/Dockerfile`.

**Comment by cmoulliard** [comment]
> would that solve this particular problem?

I can do that of course ;-)

**Comment by cmoulliard** [comment]
> pointed out we need to have some expectation of project structure even if it's not related to the command.

I agree with you that we need a project structure and such info should be reported by the command if it fails. It is not required to show the project structure that you expect to have BUT at least that the command includes a link to the web page doc ;-)

**Comment by LiliC** [comment]
> I agree with you that we need a project structure and such info should be reported by the command if it fails. It is not required to show the project structure that you expect to have BUT at least that the command includes a link to the web page doc ;-)

Not sure about the link to the web page doc in an error msg, but made the adjustment @joelanford suggested above in the PR. We can also add more clarity to the docs or descriptions to the commands themselves, that we expect the operator-sdk project structure?

**Comment by cmoulliard** [comment]
> We can also add more clarity to the docs or descriptions to the commands themselves, that we expect the operator-sdk project structure?

Yes of course. 

> the link to the web page doc in an error msg

`oc` client provides such link about where users can find more info

```
$ oc get -h
...
Usage:
  oc get
[(-o|--output=)json|yaml|wide|custom-columns=...|custom-columns-file=...|go-template=...|go-template-file=...|jsonpath=...|jsonpath-file=...]
(TYPE[.VERSION][.GROUP] [NAME | -l label] | TYPE[.VERSION][.GROUP]/NAME ...) [flags]
  -o, --output='': Output format. One of:
json|yaml|wide|name|custom-columns=...|custom-columns-file=...|go-template=...|go-template-file=...|jsonpath=...|jsonpath-file=...
See custom columns [http://kubernetes.io/docs/user-guide/kubectl-overview/#custom-columns], golang template
[http://golang.org/pkg/text/template/#pkg-overview] and jsonpath template
[http://kubernetes.io/docs/user-guide/jsonpath].
```
